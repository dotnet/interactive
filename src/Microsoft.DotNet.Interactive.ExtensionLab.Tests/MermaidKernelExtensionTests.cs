// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;

using Assent;

using FluentAssertions;
using FluentAssertions.Execution;
using HtmlAgilityPack;
using Microsoft.DotNet.Interactive.Commands;
using Microsoft.DotNet.Interactive.CSharp;
using Microsoft.DotNet.Interactive.Events;
using Microsoft.DotNet.Interactive.Formatting;
using Microsoft.DotNet.Interactive.Tests.Utility;

using Xunit;

namespace Microsoft.DotNet.Interactive.ExtensionLab.Tests
{
    public class MermaidKernelExtensionTests : IDisposable
    {
        private readonly Configuration _configuration;

        public MermaidKernelExtensionTests()
        {
            _configuration = new Configuration()
                .UsingExtension("txt")
                .SetInteractive(Debugger.IsAttached);
        }

        [Fact]
        public async Task adds_mermaid_kernel()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            kernel.ChildKernels
                .Should()
                .ContainSingle(k => k is MermaidKernel);
        }

        [Fact]
        public async Task registers_html_formatter_for_MermaidMarkdown()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            var markdown = @"graph TD
    A[Client] --> B[Load Balancer]
    B --> C[Server1]
    B --> D[Server2]";

            var formatted = new MermaidMarkdown(markdown).ToDisplayString(HtmlFormatter.MimeType);
            var doc = new HtmlDocument();
            doc.LoadHtml(formatted.FixedGuid());
            var scriptNode = doc.DocumentNode.SelectSingleNode("//div/script");
            var renderTarget = doc.DocumentNode.SelectSingleNode("//div[@id='00000000000000000000000000000000']");
            using var _ = new AssertionScope();

            scriptNode.Should().NotBeNull();
            scriptNode.InnerText.Should()
                .Contain(markdown);
            scriptNode.InnerText.Should()
                .Contain("configureRequireFromExtension('Mermaid','1.0.0')(['Mermaid/mermaidapi'], (mermaid) => {");

            renderTarget.Should().NotBeNull();
        }

        [Fact]
        public async Task registers_html_formatter_for_explorer()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            var explorer = typeof(List<string>).ExploreWithUmlClassDiagram();
            var formatted = explorer.ToDisplayString(HtmlFormatter.MimeType);
            var markdown = explorer.ToMarkdown().ToString();

            var doc = new HtmlDocument();
            doc.LoadHtml(formatted.FixedGuid());
            var scriptNode = doc.DocumentNode.SelectSingleNode("//div/script");
            var renderTarget = doc.DocumentNode.SelectSingleNode("//div[@id='00000000000000000000000000000000']");
            
            using var _ = new AssertionScope();

            scriptNode.Should().NotBeNull();
            scriptNode.InnerText.Should()
                .Contain(markdown);
            scriptNode.InnerText.Should()
                .Contain("configureRequireFromExtension('Mermaid','1.0.0')(['Mermaid/mermaidapi'], (mermaid) => {");

            renderTarget.Should().NotBeNull();
        }


        [Fact]
        public async Task mermaid_kernel_handles_SubmitCode()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            KernelCommandResult result = await kernel.SubmitCodeAsync(@"
#!mermaid
    graph TD
    A[Client] --> B[Load Balancer]
    B --> C[Server1]
    B --> D[Server2]
");

            var events = result.KernelEvents.ToSubscribedList();

            var formattedData = events
                .OfType<DisplayedValueProduced>()
                .Single()
                .FormattedValues
                .Single(fm => fm.MimeType == HtmlFormatter.MimeType)
                .Value;

            this.Assent(formattedData.FixedGuid(), _configuration);
        }

        [Fact]
        public async Task can_use_extension_methods_from_the_kernel_extension()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            await kernel.SendAsync(new SubmitCode($@"#r ""{typeof(MermaidKernelExtension).Assembly.Location}""", "csharp"));

            var result = await kernel.SendAsync( new SubmitCode(@"
using System;

typeof(List<string>).ExploreWithUmlClassDiagram().Display();
", "csharp"));

            var events = result.KernelEvents.ToSubscribedList();

            var formattedData = events
                .OfType<DisplayedValueProduced>()
                .Single()
                .FormattedValues
                .Single(fm => fm.MimeType == HtmlFormatter.MimeType)
                .Value;

            this.Assent(formattedData.FixedGuid(), _configuration);
        }

        [Fact]
        public void can_generate_class_diagram_from_type()
        {
            var diagram = typeof(JsonElement).ToClassDiagram();

            this.Assent(diagram.ToString());
        }

        [Fact]
        public void can_generate_class_diagram_from_generic_type()
        {
            var diagram = typeof(List<Dictionary<string,object>>).ToClassDiagram();

            this.Assent(diagram.ToString());
        }


        [Fact]
        public void can_generate_class_diagram_to_a_specified_depth()
        {
            var diagram = typeof(List<Dictionary<string, object>>).ToClassDiagram(new ClassDiagramConfiguration(1));

            diagram.ToString().Should()
                .NotContain(
                    "ICollection~Dictionary<String, Object>~ --|> IEnumerable~Dictionary<String, Object>~ : Inheritance");
        }

        [Fact]
        public void can_explore_a_type_with_UmlClassDiagram()
        {
            var diagram = typeof(List<Dictionary<string, object>>).ExploreWithUmlClassDiagram().ToMarkdown();

            this.Assent(diagram.ToString());
        }

        [Fact]
        public void can_configure_UmlClassDiagramExplorer()
        {
            var diagram = typeof(List<Dictionary<string, object>>).ExploreWithUmlClassDiagram()
                .WithGraphDepth(1).ToMarkdown();
            
            diagram.ToString().Should()
                .NotContain(
                    "ICollection~Dictionary<String, Object>~ --|> IEnumerable~Dictionary<String, Object>~ : Inheritance");
        }

        public void Dispose()
        {
            Formatter.ResetToDefault();
        }
    }
}