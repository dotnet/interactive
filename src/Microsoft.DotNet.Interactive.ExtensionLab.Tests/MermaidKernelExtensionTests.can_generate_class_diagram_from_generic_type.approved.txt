classDiagram

class List~Dictionary<String, Object>~
List~Dictionary<String, Object>~ : +Add(Dictionary~String, Object~ item) Void
List~Dictionary<String, Object>~ : +AddRange(IEnumerable~Dictionary<String, Object>~ collection) Void
List~Dictionary<String, Object>~ : -AddWithResize(Dictionary~String, Object~ item) Void
List~Dictionary<String, Object>~ : +AsReadOnly() ReadOnlyCollection~Dictionary<String, Object>~
List~Dictionary<String, Object>~ : +BinarySearch(Int32 index, Int32 count, Dictionary~String, Object~ item, IComparer~Dictionary<String, Object>~ comparer) Int32
List~Dictionary<String, Object>~ : +BinarySearch(Dictionary~String, Object~ item) Int32
List~Dictionary<String, Object>~ : +BinarySearch(Dictionary~String, Object~ item, IComparer~Dictionary<String, Object>~ comparer) Int32
List~Dictionary<String, Object>~ : +Clear() Void
List~Dictionary<String, Object>~ : +Contains(Dictionary~String, Object~ item) Boolean
List~Dictionary<String, Object>~ : +ConvertAll(Converter~Dictionary<String, Object>, TOutput~ converter) List~TOutput~
List~Dictionary<String, Object>~ : +CopyTo(Dictionary array) Void
List~Dictionary<String, Object>~ : +CopyTo(Int32 index, Dictionary array, Int32 arrayIndex, Int32 count) Void
List~Dictionary<String, Object>~ : +CopyTo(Dictionary array, Int32 arrayIndex) Void
List~Dictionary<String, Object>~ : +EnsureCapacity(Int32 capacity) Int32
List~Dictionary<String, Object>~ : +Exists(Predicate~Dictionary<String, Object>~ match) Boolean
List~Dictionary<String, Object>~ : +Find(Predicate~Dictionary<String, Object>~ match) Dictionary~String, Object~
List~Dictionary<String, Object>~ : +FindAll(Predicate~Dictionary<String, Object>~ match) List~Dictionary<String, Object>~
List~Dictionary<String, Object>~ : +FindIndex(Predicate~Dictionary<String, Object>~ match) Int32
List~Dictionary<String, Object>~ : +FindIndex(Int32 startIndex, Predicate~Dictionary<String, Object>~ match) Int32
List~Dictionary<String, Object>~ : +FindIndex(Int32 startIndex, Int32 count, Predicate~Dictionary<String, Object>~ match) Int32
List~Dictionary<String, Object>~ : +FindLast(Predicate~Dictionary<String, Object>~ match) Dictionary~String, Object~
List~Dictionary<String, Object>~ : +FindLastIndex(Predicate~Dictionary<String, Object>~ match) Int32
List~Dictionary<String, Object>~ : +FindLastIndex(Int32 startIndex, Predicate~Dictionary<String, Object>~ match) Int32
List~Dictionary<String, Object>~ : +FindLastIndex(Int32 startIndex, Int32 count, Predicate~Dictionary<String, Object>~ match) Int32
List~Dictionary<String, Object>~ : +ForEach(Action~Dictionary<String, Object>~ action) Void
List~Dictionary<String, Object>~ : +get_Capacity() Int32
List~Dictionary<String, Object>~ : +get_Count() Int32
List~Dictionary<String, Object>~ : +get_Item(Int32 index) Dictionary~String, Object~
List~Dictionary<String, Object>~ : +GetEnumerator() Enumerator~Dictionary<String, Object>~
List~Dictionary<String, Object>~ : +GetRange(Int32 index, Int32 count) List~Dictionary<String, Object>~
List~Dictionary<String, Object>~ : -Grow(Int32 capacity) Void
List~Dictionary<String, Object>~ : +IndexOf(Dictionary~String, Object~ item) Int32
List~Dictionary<String, Object>~ : +IndexOf(Dictionary~String, Object~ item, Int32 index) Int32
List~Dictionary<String, Object>~ : +IndexOf(Dictionary~String, Object~ item, Int32 index, Int32 count) Int32
List~Dictionary<String, Object>~ : +Insert(Int32 index, Dictionary~String, Object~ item) Void
List~Dictionary<String, Object>~ : +InsertRange(Int32 index, IEnumerable~Dictionary<String, Object>~ collection) Void
List~Dictionary<String, Object>~ : -IsCompatibleObject(Object value) Boolean$
List~Dictionary<String, Object>~ : +LastIndexOf(Dictionary~String, Object~ item) Int32
List~Dictionary<String, Object>~ : +LastIndexOf(Dictionary~String, Object~ item, Int32 index) Int32
List~Dictionary<String, Object>~ : +LastIndexOf(Dictionary~String, Object~ item, Int32 index, Int32 count) Int32
List~Dictionary<String, Object>~ : +Remove(Dictionary~String, Object~ item) Boolean
List~Dictionary<String, Object>~ : +RemoveAll(Predicate~Dictionary<String, Object>~ match) Int32
List~Dictionary<String, Object>~ : +RemoveAt(Int32 index) Void
List~Dictionary<String, Object>~ : +RemoveRange(Int32 index, Int32 count) Void
List~Dictionary<String, Object>~ : +Reverse() Void
List~Dictionary<String, Object>~ : +Reverse(Int32 index, Int32 count) Void
List~Dictionary<String, Object>~ : +set_Capacity(Int32 value) Void
List~Dictionary<String, Object>~ : +set_Item(Int32 index, Dictionary~String, Object~ value) Void
List~Dictionary<String, Object>~ : +Sort() Void
List~Dictionary<String, Object>~ : +Sort(IComparer~Dictionary<String, Object>~ comparer) Void
List~Dictionary<String, Object>~ : +Sort(Int32 index, Int32 count, IComparer~Dictionary<String, Object>~ comparer) Void
List~Dictionary<String, Object>~ : +Sort(Comparison~Dictionary<String, Object>~ comparison) Void
List~Dictionary<String, Object>~ : -System.Collections.Generic.ICollection<T>.get_IsReadOnly() Boolean
List~Dictionary<String, Object>~ : -System.Collections.Generic.IEnumerable<T>.GetEnumerator() IEnumerator~Dictionary<String, Object>~
List~Dictionary<String, Object>~ : -System.Collections.ICollection.CopyTo(Array array, Int32 arrayIndex) Void
List~Dictionary<String, Object>~ : -System.Collections.ICollection.get_IsSynchronized() Boolean
List~Dictionary<String, Object>~ : -System.Collections.ICollection.get_SyncRoot() Object
List~Dictionary<String, Object>~ : -System.Collections.IEnumerable.GetEnumerator() IEnumerator
List~Dictionary<String, Object>~ : -System.Collections.IList.Add(Object item) Int32
List~Dictionary<String, Object>~ : -System.Collections.IList.Contains(Object item) Boolean
List~Dictionary<String, Object>~ : -System.Collections.IList.get_IsFixedSize() Boolean
List~Dictionary<String, Object>~ : -System.Collections.IList.get_IsReadOnly() Boolean
List~Dictionary<String, Object>~ : -System.Collections.IList.get_Item(Int32 index) Object
List~Dictionary<String, Object>~ : -System.Collections.IList.IndexOf(Object item) Int32
List~Dictionary<String, Object>~ : -System.Collections.IList.Insert(Int32 index, Object item) Void
List~Dictionary<String, Object>~ : -System.Collections.IList.Remove(Object item) Void
List~Dictionary<String, Object>~ : -System.Collections.IList.set_Item(Int32 index, Object value) Void
List~Dictionary<String, Object>~ : +ToArray() Dictionary
List~Dictionary<String, Object>~ : +TrimExcess() Void
List~Dictionary<String, Object>~ : +TrueForAll(Predicate~Dictionary<String, Object>~ match) Boolean


